### DISCUSSION TITLE
Metadata for storage

### DISCUSSION THREAD

**Discussion by tomtomdurrant** [supabase/supabase#5479] [Questions]
Hi, 
I'm hoping this is a simple enough query. I'm trying make a load of files publicly available, but maintain the original "owner" of the file. Is there a way to store this as metadata against a file in Storage? Or would this only be able to be achieved via some lookup table. 
[This](https://firebase.google.com/docs/storage/web/file-metadata#file_metadata_properties) is the sort of thing I'm wondering is supabase supports.
Thanks

**Comment by juliuslipp** [comment]
Any updates on this?

**Reply by GaryAustin1** [reply]
There are no extra fields in the storage schema and no API for doing this.

**Reply by Srinivasan-Akash** [reply]
Hello I have built a platform using supabase by which users can upload files it will be gzipped later stored in supabase bucket and he/she gets a shareable link to download the files which redirects to my website's download page again. Now i am thinking to add OTP security but how can i send the phone number with the zip file that is been uploaded to supabase storage bucket ?? Would be great if you help me 😁

**Reply by juliuslipp** [reply]
> There are no extra fields in the storage schema and no API for doing this.

storage.objects already has a metadata field. The js client also has a setMetadata and getMetadata function on client.storage. ..., which is commented out in the source code. So I don't really agree?

**Reply by GaryAustin1** [reply]
@juliuslipp Not arguing this would not be good, but where are those calls in storage-js client?
I've searched and don't find anything.

**Reply by juliuslipp** [reply]
@GaryAustin1 https://github.com/supabase/storage-js/blob/b2e3486222126620cc65a5a1c1f967d25b4fb099/src/packages/StorageFileApi.ts#L576-L638

**Reply by GaryAustin1** [reply]
@juliuslipp Thank you.  Code was added commented out 7 months ago and the endpoint it calls “metadata” does not exist (that I can find, but that did not go well last time).   Looked at storage-api and doc https://supabase.github.io/storage-api/#/

So not sure what the plan was.

**Reply by bombillazo** [reply]
I'm pretty sure the plan (at some point) was precisely what OP and the rest are asking for: a way to store/get file metadata. Currently, it seems like we need a separate table and query when fetching files.

**Comment by rdylina** [comment]
I also would like this feature. My use case is storing captions along with photos. Firebase has this ability, and I would love to see it on Supabase.

**Comment by barbinbrad** [comment]
If `getMetaData` and `setMetaData` are too much of a problem, could we return the `owner` field with the result? Use case is displaying a "uploaded by" column in the table of files. Requires a second fetch, but not a big deal.

**Comment by bombillazo** [comment]
So far we ended up creating a custom view that joins our own File table containing the file metadata with the storage schema objects table, and we use that as our standard file table.

**Reply by barbinbrad** [reply]
I do that too, but that requires two inserts.  That’s ok I guess, but adds complexity. My thinking is, if the data is there why not send it?

**Reply by bombillazo** [reply]
Agreed, i wish the objects table had a custom metadata field.

**Comment by abdirahmn1** [comment]
Any updates on this?, the `metadata` column on `storage.objects` table seems to be `jsonb`, i don't know why this isn't exposed?

**Reply by bombillazo** [reply]
The metadata field is currently specific to S3 metadata, not User metadata. They should add another column for user defined metadata.

**Comment by semanser** [comment]
I'm facing the exact same problem. Having two tables to manage storage metadata seems to be a bit of an overkill for this problem.

@GaryAustin1 is this something that Supabase would like to have? I can implement this feature if needed (and if there is some level of confidence that this is something that you're interested in merging). Thanks!

**Reply by GaryAustin1** [reply]
I am not an employee so can't speak for Supabase.

**Comment by cipriancaba** [comment]
Supabase is so close to being amazing. Just a bit better auth, orm and storage and it would be perfect. They really need to allow us to do a better job working with their custom schemas, otherwise we just keep duplicating everything and I really don't want to be bothered with that. This is valid for both auth and storage.

**Comment by heloineto** [comment]
I'm having the same issue

**Comment by drummerjolev** [comment]
Same issue here!

**Comment by arthur-es** [comment]
same

**Comment by GaryAustin1** [comment]
You all are commenting on a almost 2 year old question with no employees in it.
Someone might start an issue in storage repository about status of the code one of the first posters linked that is still commented out I believe.  Or start an enhancement request in discussions if it does not exist.

**Comment by bombillazo** [comment]
The Storage API seems very abandoned... We've tried using it to store our config and user files in local/prod, and it is just too rudimentary and bare for use in non-trivial cases. Here are some of the issues with it:

- No capability to store/update custom file metadata, forcing you to join a custom table/view to enrich the data
- Moving or renaming a file deletes and recreates it, giving it a new ID and breaking any relational data 
- The delete functionality MUST be from the API; files can be left hanging if managed directly from the database; Supabase doesn't use triggers/webhooks themselves to solve this...
- Inconsistent Supabase client storage responses, data returned by the client and the Storage API don't always match up

I wish Supabase gave more care to this feature; we're considering using an external Files API since the Storage API isn't up to par.

**Reply by abdirahmn1** [reply]
how did you go about solving the authorization on your new preferred blob store, supabase is convenient because you can just use some good old RLS rules to write authorization policies?, which blob store did you move to and how did you tackle the authorization problem?

Your feedback is appreciated.

**Comment by bombillazo** [comment]
PS: to users who also require metadata for files, I created an issue [here](https://github.com/supabase/storage/issues/439) to track it and see if it can get traction, and move the discussion to the pertinent repo.

**Comment by BrandiATMuhkuh** [comment]
This should be solved. Here is the PR https://github.com/supabase/storage-py/pull/328

```
storage.from_(bucketName).upload(uploadPath, file, {
        "metadata": {
          "custom": "metadata",
          "second": "second",
          "third": "third",
        },
})
```
